File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/LifecycleStateMachine.java
package com.snowplowanalytics.snowplow.internal.tracker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.entity.LifecycleEntity;
import com.snowplowanalytics.snowplow.event.Background;
import com.snowplowanalytics.snowplow.event.Event;
import com.snowplowanalytics.snowplow.event.Foreground;
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson;
import com.snowplowanalytics.snowplow.tracker.InspectableEvent;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class LifecycleStateMachine implements StateMachineInterface {

    /*
     States: Visible, NotVisible
     Events: FG (Foreground), BG (Background)
     Transitions:
      - Visible (BG) NotVisible
      - NotVisible (FG) Visible
     Entity Generation:
      - Visible, NotVisible
     */

    @NonNull
    @Override
    public List<String> subscribedEventSchemasForTransitions() {
        return Arrays.asList(Background.SCHEMA, Foreground.SCHEMA);
    }

    @NonNull
    @Override
    public List<String> subscribedEventSchemasForEntitiesGeneration() {
        return Collections.singletonList("*");
    }

    @NonNull
    @Override
    public List<String> subscribedEventSchemasForPayloadUpdating() {
        return Collections.emptyList();
    }

    @Nullable
    @Override
    public State transition(@NonNull Event event, @Nullable State currentState) {
        if (event instanceof Foreground) {
            Foreground e = (Foreground) event;
            return new LifecycleState(true, e.foregroundIndex);
        }
        if (event instanceof Background) {
            Background e = (Background) event;
            return new LifecycleState(false, e.backgroundIndex);
        }
        return null;
    }

    @Nullable
    @Override
    public List<SelfDescribingJson> entities(@NonNull InspectableEvent event, @Nullable State state) {
        if (state == null) return Collections.singletonList(new LifecycleEntity(true));

        LifecycleState s = (LifecycleState) state;
        return Collections.singletonList(new LifecycleEntity(s.isForeground).index(s.index));
    }

    @Nullable
    @Override
    public Map<String, Object> payloadValues(@NonNull InspectableEvent event, @Nullable State state) {
        return null;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/LifecycleStateMachine.kt
package com.snowplowanalytics.core.tracker

import com.snowplowanalytics.snowplow.entity.LifecycleEntity
import com.snowplowanalytics.snowplow.event.Background
import com.snowplowanalytics.snowplow.event.Event
import com.snowplowanalytics.snowplow.event.Foreground
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import com.snowplowanalytics.snowplow.tracker.InspectableEvent

class LifecycleStateMachine : StateMachineInterface {
    /*
     States: Visible, NotVisible
     Events: FG (Foreground), BG (Background)
     Transitions:
      - Visible (BG) NotVisible
      - NotVisible (FG) Visible
     Entity Generation:
      - Visible, NotVisible
     */
    
    override fun subscribedEventSchemasForTransitions(): List<String> {
        return listOf(Background.schema, Foreground.schema)
    }

    override fun subscribedEventSchemasForEntitiesGeneration(): List<String> {
        return listOf("*")
    }

    override fun subscribedEventSchemasForPayloadUpdating(): List<String> {
        return emptyList()
    }

    override fun transition(event: Event, currentState: State?): State? {
        if (event is Foreground) {
            return LifecycleState(true, event.foregroundIndex)
        }
        if (event is Background) {
            return LifecycleState(false, event.backgroundIndex)
        }
        return null
    }

    override fun entities(event: InspectableEvent, state: State?): List<SelfDescribingJson> {
        if (state == null) return listOf<SelfDescribingJson>(LifecycleEntity(true))
        
        val s = state as LifecycleState
        return listOf<SelfDescribingJson>(LifecycleEntity(s.isForeground).index(s.index))
    }

    override fun payloadValues(event: InspectableEvent, state: State?): Map<String, Any>? {
        return null
    }
}
